File: app/src/main/java/io/zenandroid/greenfield/feed/FeedPresenter.java
package io.zenandroid.greenfield.feed;

import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.support.annotation.Nullable;

import com.squareup.picasso.Picasso;
import com.squareup.picasso.Target;

import java.util.Collections;

import io.reactivex.Single;
import io.zenandroid.greenfield.base.BasePresenter;
import io.zenandroid.greenfield.model.Image;
import io.zenandroid.greenfield.model.ImageListResponse;
import io.zenandroid.greenfield.service.FlickrService;

/**
 * Created by alex on 25/01/2018.
 */

public class FeedPresenter extends BasePresenter implements FeedContract.Presenter {

    private final FlickrService flickrService;
    private final FeedContract.View view;
    private FeedContract.SortCriterion criterion = FeedContract.SortCriterion.PUBLISHED;
    private String tags;
    private ImageListResponse imageListResponse;

    public FeedPresenter(FeedContract.View view, FlickrService flickrService) {
        super(view);

        this.view = view;
        this.flickrService = flickrService;
    }

    @Override
    public void subscribe() {
        fetchImages(null);
    }

    private void setImageList(ImageListResponse imageListResponse) {
        view.dismissProgressDialog();
        view.showImages(imageListResponse.getItems());
    }

    @Override
    public void onSearch(CharSequence query) {
        view.closeSearchBox();
        view.setSubTitle(query.toString());
        unsubscribe(); // make sure we're cancelling any previous searches
        imageListResponse = null;
        tags = query.toString().replaceAll("\\s+", ",");
        fetchImages(tags);
    }

    @Override
    public void onSortBy(FeedContract.SortCriterion criterion) {
        if(criterion == this.criterion) {
            return;
        }

        unsubscribe();
        this.criterion = criterion;
        fetchImages(tags);
    }

    @Override
    public void onRefresh() {
        imageListResponse = null;
        fetchImages(tags);
    }

    @Override
    public void onBrowseImage(Image image) {
        view.browseURL(image.getLink());
    }

    @Override
    public void onShareImage(Image image) {
        view.sendEmail(image.getMedia().getM());
    }

    @Override
    public void onSaveImage(Image image) {
        view.loadImageFromURL(image.getMedia().getM(), new Target() {
            @Override
            public void onBitmapLoaded(Bitmap bitmap, Picasso.LoadedFrom from) {
                view.saveBitmapToGallery(bitmap, image.getTitle(), image.getTags());
            }

            @Override
            public void onBitmapFailed(Drawable errorDrawable) {}

            @Override
            public void onPrepareLoad(Drawable placeHolderDrawable) {}
        });
    }

    private void fetchImages(@Nullable String tags) {
        view.showProgressDialog();
        final Single<ImageListResponse> source = imageListResponse != null ?
                Single.just(imageListResponse) :
                flickrService.getImageList(tags).doOnSuccess(this::storeList);
        addDisposable(source
                        .doOnSuccess(this::storeList)
                        .doOnSuccess(this::sortList)
                        .subscribe(this::setImageList, this::onError)
        );
    }

    private void sortList(ImageListResponse imageListResponse) {
        Collections.sort(imageListResponse.getItems(), (o1, o2) -> {
            switch (criterion) {
                case TAKEN:
                    if(o1.getDateTaken() == null) {
                        return -1;
                    }
                    return o1.getDateTaken().compareTo(o2.getDateTaken());
                case PUBLISHED:
                    if(o1.getPublishedDate() == null) {
                        return -1;
                    }
                    return o1.getPublishedDate().compareTo(o2.getPublishedDate());
                default:
                    return 0;
            }
        });
    }

    private void storeList(ImageListResponse imageListResponse) {
        this.imageListResponse = imageListResponse;
    }
}

File: app/src/main/java/io/zenandroid/greenfield/feed/FeedPresenter.kt
package io.zenandroid.greenfield.feed

import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import com.squareup.picasso.Picasso
import com.squareup.picasso.Target
import io.reactivex.Single
import io.zenandroid.greenfield.base.BasePresenter
import io.zenandroid.greenfield.model.Image
import io.zenandroid.greenfield.model.ImageListResponse
import io.zenandroid.greenfield.service.FlickrService

/**
 * Created by alex on 25/01/2018.
 */

class FeedPresenter(private val view: FeedContract.View, private val flickrService: FlickrService) : BasePresenter(view), FeedContract.Presenter {
    private var criterion: FeedContract.SortCriterion = FeedContract.SortCriterion.PUBLISHED
    private var tags: String? = null
    private var imageListResponse: ImageListResponse? = null

    override fun subscribe() {
        fetchImages(null)
    }

    private fun setImageList(imageListResponse: ImageListResponse) {
        view.dismissProgressDialog()
        imageListResponse.items?.let(view::showImages)
    }

    override fun onSearch(query: CharSequence) {
        view.closeSearchBox()
        view.setSubTitle(query.toString())
        unsubscribe() // make sure we're cancelling any previous searches
        imageListResponse = null
        tags = query.toString().replace("\\s+".toRegex(), ",")
        fetchImages(tags)
    }

    override fun onSortBy(criterion: FeedContract.SortCriterion) {
        if (criterion == this.criterion) {
            return
        }

        unsubscribe()
        this.criterion = criterion
        fetchImages(tags)
    }

    override fun onRefresh() {
        imageListResponse = null
        fetchImages(tags)
    }

    override fun onBrowseImage(image: Image) {
        image.link?.let(view::browseURL)
    }

    override fun onShareImage(image: Image) {
        image.media?.m?.let(view::sendEmail)
    }

    override fun onSaveImage(image: Image) {
        image.media?.m?.let {
            view.loadImageFromURL(it, object : Target {
                override fun onBitmapLoaded(bitmap: Bitmap, from: Picasso.LoadedFrom) {
                    view.saveBitmapToGallery(bitmap,
                            image.title ?: "Default Title",
                            image.tags ?: "Default Description")
                }

                override fun onBitmapFailed(errorDrawable: Drawable) {}

                override fun onPrepareLoad(placeHolderDrawable: Drawable) {}
            })
        }
    }

    private fun fetchImages(tags: String?) {
        view.showProgressDialog()
        val source = imageListResponse?.let { Single.just(it) }
                ?: flickrService.getImageList(tags)

        addDisposable(source
                .doOnSuccess(this::storeList)
                .doOnSuccess(this::sortList)
                .subscribe(this::setImageList, this::onError)
        )
    }

    private fun sortList(imageListResponse: ImageListResponse) {
        imageListResponse.items = imageListResponse.items?.sortedBy {
            when (criterion) {
                FeedContract.SortCriterion.TAKEN -> it.dateTaken
                FeedContract.SortCriterion.PUBLISHED -> it.published
            }
        }
    }

    private fun storeList(imageListResponse: ImageListResponse) {
        this.imageListResponse = imageListResponse
    }
}
